### PR TITLE
feat: Return immutable collections

### PR DESCRIPTION
**Pull request by ynsehoornenborg**
The checks rule [5@119](https://csviewer.tiobe.com/#/ruleset/rule?setid=4T_Jr6-VSX6fp6egDIhGow&status=CHECKED,UNCHECKED&tagid=6u27MkXORKaev0VNuWR_SA&ruleid=R8DUGJzYSfWdIWt3nDk7JQ) of the Philips C# Coding Standard.

### CODE REVIEW

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ReturnImmutableCollectionsAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+			NamespaceIgnoringComparer comparer = new();
+			if(type is ArrayTypeSyntax || MutableCollections.Any(m => comparer.Compare(m, typeName) == 0))
+			{
+				var loc = type.GetLocation();
+				context.ReportDiagnostic(Diagnostic.Create(Rule, loc, typeName));
+			}
+		}
+
+		private static bool IsCallableFromOutsideClass(MemberDeclarationSyntax method)
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ReturnImmutableCollectionsAnalyzer.cs:68]: Philips.CodeAnalysis.Common

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ReturnImmutableCollectionsAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+				// Private members are allowed to return mutable collections.
+				return;
+			}
+
+			var typeName = GetTypeName(type);
+
+			NamespaceIgnoringComparer comparer = new();
+			if(type is ArrayTypeSyntax || MutableCollections.Any(m => comparer.Compare(m, typeName) == 0))
+			{
+				var loc = type.GetLocation();
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ReturnImmutableCollectionsAnalyzer.cs:63]: Should we confirm with the SemanticModel here?

**Review by bcollamore — DISMISSED** [review]

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ReturnImmutableCollectionsAnalyzer.cs (hunk 2 above)
Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ReturnImmutableCollectionsAnalyzer.cs:63]: I don't think many user would create these clashing names in their namespace. On the other hand, if we do it at the very last, it doesn't hamper performance for the majority of cases.

Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ReturnImmutableCollectionsAnalyzer.cs]: Implemented

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ReturnImmutableCollectionsAnalyzer.cs (hunk 1 above)
Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ReturnImmutableCollectionsAnalyzer.cs]: Moved and created a dedicated test case for it

**Review by bcollamore — DISMISSED** [review]

**Review by bcollamore — APPROVED** [review]

**Pull request merged**